### PR TITLE
refactor: enable Python when grammar has 4 sections

### LINKED ISSUE #95: Enable JSON ASTs when 4th grammar section detected
Parent issue: https://github.com/ourPLCC/plcc/issues/52

Add an optional 4th section to grammar files. The 4th section will be populated by Python semantics. When a grammar files is written with four section, its structure is as follows:

lexical specification
%
syntactic specification
%
syntactic checks/helpers written in Java
%
semantics written in Python
Existing 3-section grammars should function normally.

When a 4th section is detected (even when empty), it will generate the Java code needed to generate JSON ASTs. Note that the 3rd section could also be empty, and it would behave the same.

### PR DESCRIPTION
**Pull request by StoneyJackson**
This is in anticipation of a new hybrid Java/Python mode in which the parser runs in Java and the semantics rep-loop runs in Python.

In this commit, when an optional 4th section is detected, plcc.py generates ParseJsonAst.json. Thus allowing the parser to to generate JSON ASTs when the `--json_ast` flag is passed. It also generates experimental Python files.

---

Closes #95

### CODE REVIEW

**Comment by StoneyJackson** [comment]
@reedeveris @AksharP5 

Yesterday, after we merged #96, @AksharP5 noticed that one of the ourPLCC/language's tests fails. I started to investigate and noticed that the new commit undid a couple of the new commits that were merged into main, including a script to run the ourPLCC/language's tests in ourPLCC/plcc.

Because of these problems, I have created a new `python` branch and it is associated with this merge request. Then I reset main to the previous merge commit, and force pushed it (so make sure you don't reuse an existing GItPod workspace; get a fresh one).

So at this point, this PR is challenged and may not get merged if we can't adequately address the problems.

Personally, I think the best approach would be to reproduce the work in this PR in a sequence of smaller PRs that start from `main`. Since you already solved the hard problems, this should be relatively "easy" and instead you can focus on making each change tight, clean, simple, and obvious. You could try a different approach, but I think this approach has the greatest chance of success.

**Comment by AksharP5** [comment]
This is the error that SET was throwing with the latest version of PLCC
image

**Comment by reedeveris** [comment]
@StoneyJackson
Please review for merge
I have fixed the problem with the SET language test, so all language tests pass now.
I have fixed the errant commits and resquashed, and retested everything after squashing and pushing, so I believe this is ready to merge again.

File: src/plcc.py [hunk 1]
```diff
@@ -668,12 +690,44 @@ def makeAbstractStub(base):
...
     }}
 
-//{base}//
+    //{base}//
 }}
 """.format(cls=cls,
            base=base,
            ext=ext,
-           cases='\n'.join(indent(2, caseList))
+           cases='\n\t\t\t'.join(caseList)
```
Comment (StoneyJackson) [inline src/plcc.py:698]: Why not use indent() as before? I'm suspicious of hardcoding tabs here when there was a nice function that added the indents before.

File: src/plcc.py [hunk 2]
```diff
@@ -668,12 +690,44 @@ def makeAbstractStub(base):
...
+
+class {base}({ext}): #{base}:class#
+
+    className = "{base}"
+
+    #{base}#
+""".format(cls=cls,
+           base=base,
+           ext=ext,
+           cases='\n\t\t\t'.join(caseList)
```
Comment (StoneyJackson) [inline src/plcc.py:730]: Ditto.

File: src/plcc.py [hunk 3]
```diff
@@ -724,31 +778,92 @@ def makeStub(cls):
...
 
-//{cls}//
+    //{cls}//
 }}
 """.format(cls=cls,
            lhs=lhs,
            ext=ext,
            ruleString=ruleString,
-           decls='\n'.join(indent(1, decls)),
+           decls='\t'.join(decls),
```
Comment (StoneyJackson) [inline src/plcc.py:800]: ditto

File: src/plcc.py [hunk 4]
```diff
@@ -724,31 +778,92 @@ def makeStub(cls):
...
            params=', '.join(params),
-           inits='\n'.join(indent(2, inits)),
+           inits='\n\t\t'.join(inits),
```
Comment (StoneyJackson) [inline src/plcc.py:802]: ditto

File: src/plcc.py [hunk 5]
```diff
@@ -724,31 +778,92 @@ def makeStub(cls):
...
+    def __init__({params}):
+        #{cls}:init#
+        {inits}
+
+    #{cls}#
+    """.format(cls=cls,
+            lhs=lhs,
+            ext=ext,
+            ruleString=ruleString,
+            decls='\n\t'.join(decls),
```
Comment (StoneyJackson) [inline src/plcc.py:861]: ditto

File: src/plcc.py [hunk 6]
```diff
@@ -724,31 +778,92 @@ def makeStub(cls):
...
+            params=', '.join(params),
+            inits='\n\t\t'.join(inits),
```
Comment (StoneyJackson) [inline src/plcc.py:863]: ditto

File: src/plcc.py [hunk 7]
```diff
@@ -935,6 +1062,13 @@ def sem(nxt):
             continue
         # check to see if line has the form Class:mod
         mod = mod.strip() # mod might be 'import', 'top', etc.
+        if cls in stubs:
+            if mod and mod != 'ignore' and mod != 'top':
+                codeString = '\n\t\t'.join(codeString)
```
Comment (StoneyJackson) [inline src/plcc.py:1067]: ditto

File: src/plcc.py [hunk 8]
```diff
@@ -935,6 +1062,13 @@ def sem(nxt):
...
+            else:
+                codeString = '\n\t'.join(codeString)
```
Comment (StoneyJackson) [inline src/plcc.py:1069]: ditto

File: src/plcc.py [hunk 9]
```diff
@@ -959,18 +1093,19 @@ def sem(nxt):
                 stub = stub.replace(repl, '{} {}'.format(codeString,repl))
             else: # the default
                 repl = '//{}//'.format(cls)
-                stub = stub.replace(repl, '{}\n\n{}'.format(codeString,repl))
+                stub = stub.replace(repl, '{}\n\n\t{}'.format(codeString,repl))
```
Comment (StoneyJackson) [inline src/plcc.py:1096]: ditto

File: src/plcc.py [hunk 10]
```diff
@@ -1014,6 +1156,100 @@ def semFinishUp():
...
+        # print('>>> cls={} mod={}'.format(cls, mod))
+        cls = cls.strip()
+        codeString = getCode(nxt) # grab the stuff between %%% ... %%%
+        if line[-8:] == ':ignore!':
+            continue
+        # check to see if line has the form Class:mod
+        mod = mod.strip() # mod might be 'import', 'top', etc.
+        if cls in python_stubs:
+            if mod and mod != 'ignore' and mod != 'top':
+                codeString = '\n\t\t'.join(codeString)
```
Comment (StoneyJackson) [inline src/plcc.py:1181]: ditto

File: src/plcc.py [hunk 11]
```diff
@@ -1014,6 +1156,100 @@ def semFinishUp():
...
+            else:
+                codeString = '\n\t'.join(codeString)
```
Comment (StoneyJackson) [inline src/plcc.py:1183]: ditto

File: src/plcc.py [hunk 12]
```diff
@@ -1014,6 +1156,100 @@ def semFinishUp():
...
+        if cls in python_stubs:
+            stub = python_stubs[cls]
+            if mod:
+                repl = '#{}:{}#'.format(cls, mod)
+                stub = stub.replace(repl, '{}\n{}'.format(codeString,repl))
+                repl = "'''{}:{}'''".format(cls, mod)
+                stub = stub.replace(repl, '{} {}'.format(codeString,repl))
+            else: # the default
+                repl = '#{}#'.format(cls)
+                stub = stub.replace(repl, '{}\n\n\t{}'.format(codeString,repl))
```
Comment (StoneyJackson) [inline src/plcc.py:1210]: ditto

**Review by StoneyJackson — COMMENTED** [review]
My main concern is hard-coding indentation rather than using the provided indent function. If we decide that we later want to use spaces instead of tabs, we'll have to track down every line with \t, assess whether that \t represents indentation or has some other purpose, and then make the change.

**Comment by StoneyJackson** [comment]
Good work. Thanks @reedeveris @AksharP5 !

**Pull request merged**